packages/core/src/modules/dids/methods/peer/peerDidNumAlgo4.ts

```diff
@@ -0,0 +1,126 @@
+export const isLongFormDidPeer4 = (did: string) => LONG_RE.test(did)
+
+const hashEncodedDocument = (encodedDocument: string) =>
+  MultiBaseEncoder.encode(
+    MultiHashEncoder.encode(TypedArrayEncoder.fromString(encodedDocument), 'sha2-256'),
+    'base58btc'
+  )
+
+export function didToNumAlgo4DidDocument(did: string) {
+  const parsed = parseDid(did)
+
+  const match = parsed.did.match(LONG_RE)
+  if (!match) {
+    throw new AriesFrameworkError(`Invalid long form algo 4 did:peer: ${parsed.did}`)
+  }
+  const [, hash, encodedDocument] = match
+  if (hash !== hashEncodedDocument(encodedDocument)) {
+    throw new AriesFrameworkError(`Hash is invalid for did: ${did}`)
+  }
+
+  const { data } = MultiBaseEncoder.decode(encodedDocument)
+  const [multiCodecValue] = VarintEncoder.decode(data.subarray(0, 2))
+  if (multiCodecValue !== 0x0200) {
+    throw new AriesFrameworkError(`Not a JSON multicodec`)
+  }
+  const didDocumentJson = JsonEncoder.fromBuffer(data.subarray(2))
+
+  didDocumentJson.id = parsed.did
+  didDocumentJson.alsoKnownAs = [parsed.did.slice(0, did.lastIndexOf(':'))]
+
+  // Populate all verification methods without controller
+  const addControllerIfNotPresent = (item: unknown) => {
+    if (Array.isArray(item)) item.forEach(addControllerIfNotPresent)
+
+    if (item && typeof item === 'object' && (item as Record<string, unknown>).controller === undefined) {
+      ;(item as Record<string, unknown>).controller = parsed.did
+    }
+  }
+
+  addControllerIfNotPresent(didDocumentJson.verificationMethod)
+  addControllerIfNotPresent(didDocumentJson.authentication)
+  addControllerIfNotPresent(didDocumentJson.assertionMethod)
+  addControllerIfNotPresent(didDocumentJson.keyAgreement)
+  addControllerIfNotPresent(didDocumentJson.capabilityDelegation)
+  addControllerIfNotPresent(didDocumentJson.capabilityInvocation)
+
+  const didDocument = JsonTransformer.fromJSON(didDocumentJson, DidDocument)
+  return didDocument
+}
+
+export function didDocumentToNumAlgo4Did(didDocument: DidDocument) {
+  const didDocumentJson = didDocument.toJSON()
+
+  const deleteControllerIfPresent = (item: unknown) => {
+    if (Array.isArray(item)) {
+      item.forEach((method: { controller?: string }) => {
+        if (method.controller === '#id' || method.controller === didDocument.id) delete method.controller
+      })
+    }
+  }
+  delete didDocumentJson.id
+  delete didDocumentJson.alsoKnownAs
+  deleteControllerIfPresent(didDocumentJson.verificationMethod)
+  deleteControllerIfPresent(didDocumentJson.authentication)
+  deleteControllerIfPresent(didDocumentJson.assertionMethod)
+  deleteControllerIfPresent(didDocumentJson.keyAgreement)
+  deleteControllerIfPresent(didDocumentJson.capabilityDelegation)
+  deleteControllerIfPresent(didDocumentJson.capabilityInvocation)
+
+  const buffer = Buffer.concat([VarintEncoder.encode(0x0200), Buffer.from(JSON.stringify(didDocumentJson))])
```

Comment:
What does 0x0200 stand for? I think if we can assign it to a variable, it makes the code a bit easier to read

Reply:
Does this rely on a consistent json stringify?

Reply:
0x0200 is the multicodec prefix for JSON (see the procedure we follow in https://identity.foundation/peer-did-method-spec/#creating-a-did).

Regarding the JSON stringify, it's something I was wondering when I've first read the spec but I don't find a problem as long as the process for a given did is only called once, and then stored for further interactions. Do you foresee issues due to this?

Reply:
Well this was a problem with did:peer:1 as well, and the whole reason json canonicalization for example exists.

If you send over the data already encoded though it's fine (e.g. with a jwt the content is included beteen the first and second dot). I need to look a bit better at did:peer:4 spec

Reply:
I've been lurking on this PR lol; the data is sent already encoded. The verification of the hash over the doc is performed by taking the hash over the still encoded doc, much like the way jwts are verified.